File: js/src/rtiruntimecontrol.ts

// Utility class to simplify sending and responding to "start", "stop" messages etc.
// and to add fast-time worker support to a TypeScript/JavaScript RTI client.

import { RTIClient, DispatchMode } from "./rticlient.js"
import { channel as RTIchannel, capability as RTIcapability } from "./constants.js"
import { RuntimeControl, RuntimeControl_LoadScenario, RuntimeControl_TimeSync } from "./generated/RuntimeControl.js"
import { FastTimeControl } from "./generated/FastTimeControl.js"
import { RuntimeState } from "./generated/RuntimeState.js"
import { Clients } from "./generated/Clients.js"
import { Log } from "./generated/Logs.js"

export class StepGrant {
    readonly stepNumber: number
    readonly startTime: number
    readonly endTime: number
    readonly timeStep: number
    /** @internal */ readonly _runId: string
    /** @internal */ readonly _realStart: number

    /** @internal */
    constructor(proto: { stepNumber: number; startTime: number; endTime: number }, runId: string) {
        this.stepNumber = proto.stepNumber
        this.startTime = proto.startTime
        this.endTime = proto.endTime
        this.timeStep = proto.endTime - proto.startTime
        this._runId = runId
        this._realStart = Date.now()
    }
}

export class RTIRuntimeControl {
    protected readonly rti: RTIClient

    scenario: RuntimeControl_LoadScenario | undefined = undefined
    publishScenario: boolean = false
    asyncReady: boolean = false
    timeScale: number | undefined = undefined
    currentLog: Log | undefined = undefined

    private _subscribed: boolean = false
    private readonly _stepFn: ((grant: StepGrant) => void | Promise<void>) | undefined
    private readonly _fastTimeEnabled: boolean
    private _fastTimeRunId: string | undefined = undefined
    private _fastTimeControllerClientId: string | undefined = undefined

    // Promise-based grant queue
    private _grantWaiters: Array<(grant: StepGrant | null) => void> = []
    private _grantQueue: StepGrant[] = []

    get isFastTime(): boolean {
        return this._fastTimeRunId !== undefined
    }

    constructor(rti: RTIClient, subscribe = true, fastTime = false, stepFn?: (grant: StepGrant) => void | Promise<void>) {
        this.rti = rti
        this._stepFn = stepFn
        this._fastTimeEnabled = fastTime || stepFn !== undefined

        if (!rti.capabilities.includes(RTIcapability.runtimeControl)) rti.capabilities.push(RTIcapability.runtimeControl)
        if (!rti.capabilities.includes(RTIcapability.scenario)) rti.capabilities.push(RTIcapability.scenario)
        if (!rti.capabilities.includes(RTIcapability.timeScale)) rti.capabilities.push(RTIcapability.timeScale)

        rti.state = RuntimeState.INITIAL

        if (this._fastTimeEnabled) {
            if (!rti.capabilities.includes(RTIcapability.fastTimeWorker))
                rti.capabilities.push(RTIcapability.fastTimeWorker)
        }

        if (subscribe) this.subscribe()
    }

    // Override hooks — subclass and override to add custom behavior

    onReset(): void {}
    onLoadScenario(loadScenario: RuntimeControl_LoadScenario, playback: boolean): boolean { return true }
    onStart(): void {}
    onPlay(): void {}
    onPause(): void {}
    onEnd(): void {}
    onStop(): void {}
    onEndStop(): void {}
    onResetEndStop(): void {}
    onTimeScale(timeScale: number): void {}
    onTimeSync(timeSync: RuntimeControl_TimeSync): void {}

    /** Called when a fast-time step grant is received. Override to add custom behavior.
     * Only called when using the getStepGrant() pattern (no stepFn provided). */
    onStepGrant(grant: StepGrant): void {}

    /** Returns a Promise that resolves to a StepGrant when one arrives, or null on timeout/stop.
     * For use in an async loop. Call completeStep(grant) when simulation work is done.
     *
     * In Node.js (the primary target), the event loop handles incoming messages while the
     * caller is awaiting this promise, so a blocking timeout is safe and natural:
     *   while (true) { const grant = await runtime.getStepGrant(); if (!grant) continue; ... }
     *
     * @param timeoutMs Timeout in milliseconds. 0 = no timeout (wait indefinitely). Default: 1000.
     */
    getStepGrant(timeoutMs: number = 1000): Promise<StepGrant | null> {
        if (this._grantQueue.length > 0) {
            return Promise.resolve(this._grantQueue.shift()!)
        }
        return new Promise<StepGrant | null>((resolve) => {
            let settled = false
            const settle = (value: StepGrant | null) => {
                if (!settled) {
                    settled = true
                    const idx = this._grantWaiters.indexOf(settle)
                    if (idx >= 0) this._grantWaiters.splice(idx, 1)
                    resolve(value)
                }
            }
            this._grantWaiters.push(settle)
            if (timeoutMs > 0) setTimeout(() => settle(null), timeoutMs)
        })
    }

    /** Send StepComplete to the fast-time controller. */
    completeStep(grant: StepGrant, failed = false, reason = ""): void {
        const duration = Date.now() - grant._realStart
        this.rti.publish(RTIchannel.fastTimeControl, FastTimeControl, {
            stepComplete: {
                clientId: this.rti.clientId,
                runId: grant._runId,
                stepNumber: grant.stepNumber,
                duration,
                failed,
                reason,
            }
        }, false)
    }

    // Runtime control publish methods

    reset(): void { this._publishAndReceive({ reset: {} }) }
    loadScenario(scenarioName: string): void { this._publishAndReceive({ loadScenario: { name: scenarioName, parameterValues: {} } }) }
    start(): void { this._publishAndReceive({ start: {} }) }
    play(): void { this._publishAndReceive({ play: {} }) }
    pause(): void { this._publishAndReceive({ pause: {} }) }
    end(): void { this._publishAndReceive({ end: {} }) }
    stop(): void { this._publishAndReceive({ stop: {} }) }
    setTimeScale(timeScale: number): void { this._publishAndReceive({ setTimeScale: { timeScale } }) }
    seek(time: number): void { this._publishAndReceive({ seek: { time } }) }
    requestCurrentLog(): void { this._publishAndReceive({ requestCurrentLog: {} }) }

    subscribe(): void {
        if (!this._subscribed) {
            const onMsg = (_ch: string, message: RuntimeControl) => this._receive(message)
            // Always IMMEDIATE so stop/end/reset pierce BUFFERED mode during fast-time steps
            this.rti.subscribe(RTIchannel.control, RuntimeControl, onMsg, false, DispatchMode.IMMEDIATE)
            // Subscribe to own (targeted) channel once clientId is known
            const subscribeOwn = () => {
                this.rti.subscribe(this.rti.ownChannelPrefix + RTIchannel.control, RuntimeControl, onMsg, false, DispatchMode.IMMEDIATE)
            }
            if (this.rti.clientId) subscribeOwn()
            else this.rti.once("firstconnect", subscribeOwn)
            if (this._fastTimeEnabled) {
                this.rti.subscribe(RTIchannel.fastTimeControl, FastTimeControl,
                    (_ch: string, message: FastTimeControl) => this._receiveFastTime(message),
                    false, DispatchMode.IMMEDIATE)
                this.rti.subscribeText(RTIchannel.clientDisconnect,
                    (_ch: string, clientId: string) => this._onControllerDisconnect(clientId),
                    false, DispatchMode.IMMEDIATE)
            }
            this._subscribed = true
        }
    }

    /** Async wait until all clients with the given application are in one of the specified states.
     * Throws on timeout. */
    async waitForApplicationState(application: string, states: RuntimeState | RuntimeState[], timeoutSec: number = 30): Promise<void> {
        if (!this._subscribed) throw new Error("Cannot wait for application state without being subscribed")
        const stateSet = new Set(Array.isArray(states) ? states : [states])
        if (!this.rti.knownClients.some(c => c.application === application)) {
            this.rti.publish(RTIchannel.clients, Clients, { requestClients: {} }, false)
        }
        const deadline = Date.now() + timeoutSec * 1000
        while (true) {
            const clients = this.rti.knownClients.filter(c => c.application === application)
            if (clients.length > 0 && clients.every(c => stateSet.has(c.state!))) return
            if (Date.now() > deadline) throw new Error(`Timeout waiting for ${application} state ${[...stateSet].join(", ")}`)
            await sleep(100)
        }
    }

    /** Async wait until the client with the given ID is in one of the specified states.
     * Throws on timeout. */
    async waitForClientState(clientId: string, states: RuntimeState | RuntimeState[], timeoutSec: number = 30): Promise<void> {
        if (!this._subscribed) throw new Error("Cannot wait for client state without being subscribed")
        const stateSet = new Set(Array.isArray(states) ? states : [states])
        if (!this.rti.knownClients.some(c => c.id === clientId)) {
            this.rti.publish(RTIchannel.clients, Clients, { requestClients: {} }, false)
        }
        const deadline = Date.now() + timeoutSec * 1000
        while (true) {
            const client = this.rti.knownClients.find(c => c.id === clientId)
            if (client && stateSet.has(client.state!)) return
            if (Date.now() > deadline) throw new Error(`Timeout waiting for client ${clientId} state ${[...stateSet].join(", ")}`)
            await sleep(100)
        }
    }

    private _publishAndReceive(message: Partial<RuntimeControl>): void {
        this.rti.publish(RTIchannel.control, RuntimeControl, message, false)
        if (!this.rti.isConnected || !this._subscribed) this._receive(RuntimeControl.fromPartial(message))
    }

    private _onControllerDisconnect(clientId: string): void {
        if (this._fastTimeControllerClientId === clientId && this.isFastTime &&
            this.rti.state !== RuntimeState.RUNNING && this.rti.state !== RuntimeState.PAUSED) {
            this._resetFastTime()
        }
    }

    private _receiveFastTime(message: FastTimeControl): void {
        if (message.configure) {
            this._fastTimeRunId = message.configure.runId
            this._fastTimeControllerClientId = message.configure.controllerClientId
            // defaultDispatchMode stays IMMEDIATE until the first step grant arrives
            this.rti.publish(RTIchannel.fastTimeControl, FastTimeControl, {
                acknowledge: {
                    clientId: this.rti.clientId,
                    runId: message.configure.runId,
                    failed: false,
                    reason: "",
                }
            }, false)
            this.rti.fastTimeMode = true
        } else if (message.stepGrant && message.stepGrant.runId === this._fastTimeRunId) {
            const grant = new StepGrant(message.stepGrant, this._fastTimeRunId!)
            this.rti.defaultDispatchMode = DispatchMode.BUFFERED // switch to BUFFERED on first step
            this.rti.flushBuffers() // dispatch messages buffered since last step
            if (this._stepFn) {
                Promise.resolve(this._stepFn(grant))
                    .then(() => this.completeStep(grant))
                    .catch((e: any) => this.completeStep(grant, true, e?.message ?? String(e)))
            } else {
                this.onStepGrant(grant)
                if (this._grantWaiters.length > 0) {
                    this._grantWaiters.shift()!(grant)
                } else {
                    this._grantQueue.push(grant)
                }
            }
        }
    }

    private _resetFastTime(): void {
        if (this._fastTimeRunId !== undefined) {
            this._fastTimeRunId = undefined
            this._fastTimeControllerClientId = undefined
            // Cancel all pending waiters and drain the queue
            const waiters = this._grantWaiters.splice(0)
            this._grantQueue.length = 0
            for (const waiter of waiters) waiter(null)
            this.rti.defaultDispatchMode = DispatchMode.IMMEDIATE
            this.rti.flushBuffers()
            this.rti.fastTimeMode = false
        }
    }

    private _receive(message: RuntimeControl): void {
        if (message.reset !== undefined) {
            this.onResetEndStop()
            this.onReset()
            this.rti.state = RuntimeState.INITIAL
            if (this._fastTimeEnabled) this._resetFastTime()
        } else if (message.loadScenario !== undefined) {
            this.scenario = undefined
            const playback = this.rti.state === RuntimeState.PLAYBACK
            this.rti.state = RuntimeState.LOADING
            const success = this.onLoadScenario(message.loadScenario, playback)
            if (!success) {
                this.rti.state = RuntimeState.UNKNOWN
                return
            }
            this.scenario = message.loadScenario
            this.rti.state = playback ? RuntimeState.PLAYBACK : RuntimeState.READY
        } else if (message.requestCurrentScenario !== undefined && this.publishScenario && this.scenario) {
            this.rti.publish(RTIchannel.control, RuntimeControl, {
                currentScenario: { name: this.scenario.name, parameterValues: this.scenario.parameterValues }
            }, false)
        } else if (message.start !== undefined) {
            this.onStart()
            this.rti.state = RuntimeState.RUNNING
        } else if (message.play !== undefined) {
            this.onPlay()
            this.rti.state = RuntimeState.PLAYBACK
            if (this._fastTimeEnabled) this._resetFastTime()
        } else if (message.pause !== undefined) {
            this.onPause()
            if (this.rti.state === RuntimeState.PLAYBACK || this.rti.state === RuntimeState.PLAYBACK_PAUSED)
                this.rti.state = RuntimeState.PLAYBACK_PAUSED
            else if (
                this.rti.state !== RuntimeState.END && this.rti.state !== RuntimeState.PLAYBACK_END &&
                this.rti.state !== RuntimeState.STOPPED && this.rti.state !== RuntimeState.PLAYBACK_STOPPED
            )
                this.rti.state = RuntimeState.PAUSED
        } else if (message.end !== undefined) {
            this.onResetEndStop()
            this.onEndStop()
            this.onEnd()
            this.rti.state = this.rti.state === RuntimeState.PLAYBACK ? RuntimeState.PLAYBACK_END : RuntimeState.END
            if (this._fastTimeEnabled) this._resetFastTime()
        } else if (message.stop !== undefined) {
            this.onResetEndStop()
            this.onEndStop()
            this.onStop()
            if (
                this.rti.state === RuntimeState.PLAYBACK || this.rti.state === RuntimeState.PLAYBACK_PAUSED ||
                this.rti.state === RuntimeState.PLAYBACK_STOPPED || this.rti.state === RuntimeState.PLAYBACK_END
            )
                this.rti.state = RuntimeState.PLAYBACK_STOPPED
            else
                this.rti.state = RuntimeState.STOPPED
            if (this._fastTimeEnabled) this._resetFastTime()
        } else if (message.setTimeScale !== undefined) {
            this.timeScale = message.setTimeScale.timeScale
            this.onTimeScale(message.setTimeScale.timeScale)
        } else if (message.timeSync !== undefined) {
            this.timeScale = message.timeSync.timeScale
            this.onTimeSync(message.timeSync)
        } else if (message.currentLog !== undefined) {
            this.currentLog = message.currentLog
        } else if (message.currentScenario !== undefined) {
            this.scenario = message.currentScenario
        }
    }
}

function sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms))
}

File: dotnet/src/RTIRuntimeControl.cs

// Utility class to simplify sending and responding to runtime control messages (start, stop, etc.)
// and to add fast-time worker support to a C# RTI client application.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Collections.Concurrent;
using Inhumate.RTI.Proto;

namespace Inhumate.RTI {

    public class StepGrant {
        public long StepNumber { get; }
        public double StartTime { get; }
        public double EndTime { get; }
        public double TimeStep { get; }
        internal string RunId { get; }
        internal DateTime RealStart { get; }

        internal StepGrant(FastTimeControl.Types.StepGrant proto, string runId) {
            StepNumber = proto.StepNumber;
            StartTime = proto.StartTime;
            EndTime = proto.EndTime;
            TimeStep = proto.EndTime - proto.StartTime;
            RunId = runId;
            RealStart = DateTime.Now;
        }
    }

    public class RTIRuntimeControl {

        protected readonly RTIClient rti;

        private bool subscribed;
        private readonly Action<StepGrant> stepFn;
        private readonly bool fastTimeEnabled;
        private string fastTimeRunId;
        private string fastTimeControllerClientId;
        private BlockingCollection<StepGrant> grantQueue = new BlockingCollection<StepGrant>(new ConcurrentQueue<StepGrant>());
        private CancellationTokenSource resetCts = new CancellationTokenSource();

        public double? TimeScale { get; private set; }
        public RuntimeControl.Types.ScenarioSpecification Scenario { get; private set; }

        public bool IsFastTime => fastTimeRunId != null;

        public RTIRuntimeControl(RTIClient rti, bool subscribe = true, bool fastTime = false, Action<StepGrant> stepFn = null) {
            this.rti = rti;
            this.stepFn = stepFn;
            this.fastTimeEnabled = fastTime || stepFn != null;

            rti.Capabilities.Add(RTICapability.RuntimeControl);
            rti.Capabilities.Add(RTICapability.Scenario);
            rti.Capabilities.Add(RTICapability.TimeScale);

            rti.State = RuntimeState.Initial;
            if (fastTimeEnabled) rti.Capabilities.Add(RTICapability.FastTimeWorker);

            if (subscribe) Subscribe();
        }

        // Virtual override hooks — override these in a subclass to add custom behavior.

        public virtual void OnReset() {}
        public virtual bool OnLoadScenario(RuntimeControl.Types.ScenarioSpecification loadScenario, bool playback) => true;
        public virtual void OnStart() {}
        public virtual void OnPlay() {}
        public virtual void OnPause() {}
        public virtual void OnEnd() {}
        public virtual void OnStop() {}
        public virtual void OnEndStop() {}
        public virtual void OnResetEndStop() {}
        public virtual void OnTimeScale(double timeScale) {}
        public virtual void OnTimeSync(RuntimeControl.Types.TimeSync timeSync) {}
        public virtual void OnSeek(RuntimeControl.Types.Seek seek) {}

        /// Called when a fast-time step grant is received. Override to add custom behavior.
        /// Only called when using the GetStepGrant() pattern (no stepFn provided).
        public virtual void OnStepGrant(StepGrant grant) {}

        /// Block until a fast-time step grant arrives. Returns a StepGrant or null on timeout/stop.
        /// For use in a polling loop. Call CompleteStep(grant) when simulation work is done.
        ///
        /// IMPORTANT: In polling mode (rti.Polling = true), use timeout = 0 so that GetStepGrant
        /// returns immediately when no grant is queued, allowing Poll() to read the incoming StepGrant
        /// from the socket. A blocking timeout would stall Poll() and delay grant delivery.
        /// In multi-threaded mode (the default), a blocking timeout is fine.
        public StepGrant GetStepGrant(int timeout = 1) {
            var timeoutMs = timeout <= 0 ? 0 : timeout;
            try {
                if (grantQueue.TryTake(out var grant, timeoutMs, resetCts.Token))
                    return grant;
                return null;
            } catch (OperationCanceledException) {
                return null;
            }
        }

        /// Send StepComplete to the fast-time controller
        public void CompleteStep(StepGrant grant, bool failed = false, string reason = "") {
            var duration = (int)(DateTime.Now - grant.RealStart).TotalMilliseconds;
            var msg = new FastTimeControl {
                StepComplete = new FastTimeControl.Types.StepComplete {
                    ClientId = rti.ClientId,
                    RunId = grant.RunId,
                    StepNumber = grant.StepNumber,
                    Duration = duration,
                }
            };
            if (failed) {
                msg.StepComplete.Failed = true;
                if (!string.IsNullOrEmpty(reason)) msg.StepComplete.Reason = reason;
            }
            rti.Publish(RTIChannel.FastTimeControl, msg);
        }

        // Runtime control publish methods

        public void Reset() => PublishAndReceive(new RuntimeControl { Reset = new Google.Protobuf.WellKnownTypes.Empty() });
        public void LoadScenario(string scenarioName) => PublishAndReceive(new RuntimeControl { LoadScenario = new RuntimeControl.Types.ScenarioSpecification { Name = scenarioName } });
        public void Start() => PublishAndReceive(new RuntimeControl { Start = new Google.Protobuf.WellKnownTypes.Empty() });
        public void Play() => PublishAndReceive(new RuntimeControl { Play = new Google.Protobuf.WellKnownTypes.Empty() });
        public void Pause() => PublishAndReceive(new RuntimeControl { Pause = new Google.Protobuf.WellKnownTypes.Empty() });
        public void End() => PublishAndReceive(new RuntimeControl { End = new Google.Protobuf.WellKnownTypes.Empty() });
        public void Stop() => PublishAndReceive(new RuntimeControl { Stop = new Google.Protobuf.WellKnownTypes.Empty() });
        public void SetTimeScale(double timeScale) => PublishAndReceive(new RuntimeControl { SetTimeScale = new RuntimeControl.Types.SetTimeScale { TimeScale = timeScale } });
        public void Seek(double time) => PublishAndReceive(new RuntimeControl { Seek = new RuntimeControl.Types.Seek { Time = time } });

        public void Subscribe() {
            if (!subscribed) {
                // Always IMMEDIATE so stop/end/reset pierce BUFFERED mode during fast-time steps
                rti.Subscribe<RuntimeControl>(RTIChannel.RuntimeControl, OnRuntimeControlMessage, dispatchMode: DispatchMode.Immediate);
                rti.Subscribe<RuntimeControl>(rti.OwnChannelPrefix + RTIChannel.RuntimeControl, OnRuntimeControlMessage, dispatchMode: DispatchMode.Immediate);
                if (fastTimeEnabled) {
                    rti.Subscribe<FastTimeControl>(RTIChannel.FastTimeControl, OnFastTimeControlMessage, dispatchMode: DispatchMode.Immediate);
                    rti.Subscribe(RTIChannel.ClientDisconnect, OnClientDisconnectMessage, false, dispatchMode: DispatchMode.Immediate);
                }
                subscribed = true;
            }
        }

        /// Block until all clients with the given application name are in one of the specified states.
        /// Throws TimeoutException if no matching clients have the desired state before the timeout.
        public void WaitForApplicationState(string application, RuntimeState state, double timeout = 30)
            => WaitForApplicationState(application, new[] { state }, timeout);

        public void WaitForApplicationState(string application, IEnumerable<RuntimeState> states, double timeout = 30) {
            if (!subscribed) throw new InvalidOperationException("Cannot wait for application state without being subscribed");
            var stateSet = new HashSet<RuntimeState>(states);
            var clients = rti.KnownClients.Where(c => c.Application == application).ToList();
            if (!clients.Any())
                rti.Publish(RTIChannel.Clients, new Clients { RequestClients = new Google.Protobuf.WellKnownTypes.Empty() });
            var deadline = DateTime.Now.AddSeconds(timeout);
            while (true) {
                clients = rti.KnownClients.Where(c => c.Application == application).ToList();
                if (clients.Any() && clients.All(c => stateSet.Contains(c.State))) return;
                if (DateTime.Now > deadline) throw new TimeoutException($"Timeout waiting for {application} state {string.Join(", ", stateSet)}");
                Thread.Sleep(100);
            }
        }

        /// Block until the client with the given ID is in one of the specified states.
        public void WaitForClientState(string clientId, RuntimeState state, double timeout = 30)
            => WaitForClientState(clientId, new[] { state }, timeout);

        public void WaitForClientState(string clientId, IEnumerable<RuntimeState> states, double timeout = 30) {
            if (!subscribed) throw new InvalidOperationException("Cannot wait for client state without being subscribed");
            var stateSet = new HashSet<RuntimeState>(states);
            if (!rti.KnownClients.Any(c => c.Id == clientId))
                rti.Publish(RTIChannel.Clients, new Clients { RequestClients = new Google.Protobuf.WellKnownTypes.Empty() });
            var deadline = DateTime.Now.AddSeconds(timeout);
            while (true) {
                var client = rti.KnownClients.FirstOrDefault(c => c.Id == clientId);
                if (client != null && stateSet.Contains(client.State)) return;
                if (DateTime.Now > deadline) throw new TimeoutException($"Timeout waiting for client {clientId} state {string.Join(", ", stateSet)}");
                Thread.Sleep(100);
            }
        }

        private void OnRuntimeControlMessage(string channel, RuntimeControl message) => Receive(message);
        private void OnFastTimeControlMessage(string channel, FastTimeControl message) => ReceiveFastTime(message);
        private void OnClientDisconnectMessage(string channel, object message) {
            var clientId = message?.ToString();
            if (clientId != null && clientId == fastTimeControllerClientId && IsFastTime &&
                rti.State != RuntimeState.Running && rti.State != RuntimeState.Paused) {
                ResetFastTime();
            }
        }

        private void PublishAndReceive(RuntimeControl message) {
            rti.Publish(RTIChannel.RuntimeControl, message);
            if (!rti.IsConnected || !subscribed) Receive(message);
        }

        private void ReceiveFastTime(FastTimeControl message) {
            switch (message.ControlCase) {
                case FastTimeControl.ControlOneofCase.ConfigureRun:
                    fastTimeRunId = message.ConfigureRun.RunId;
                    fastTimeControllerClientId = message.ConfigureRun.ControllerClientId;
                    // DefaultDispatchMode stays Immediate until the first step grant arrives
                    rti.Publish(RTIChannel.FastTimeControl, new FastTimeControl {
                        AcknowledgeRun = new FastTimeControl.Types.AcknowledgeRun {
                            ClientId = rti.ClientId,
                            RunId = message.ConfigureRun.RunId,
                        }
                    });
                    rti.FastTimeMode = true;
                    break;
                case FastTimeControl.ControlOneofCase.Configuration:
                    // A configuration with real-time (or unknown) mode means this run is not
                    // fast-time stepped — leave fast-time mode and clear the run id.
                    if (message.Configuration.Mode <= FastTimeControl.Types.ExecutionMode.RealTime) {
                        ResetFastTime();
                    }
                    break;
                case FastTimeControl.ControlOneofCase.AbandonRun:
                    // The controller abandoned the run we're configured for — leave fast-time mode.
                    if (message.AbandonRun.RunId == fastTimeRunId) {
                        ResetFastTime();
                    }
                    break;
                case FastTimeControl.ControlOneofCase.StepGrant:
                    if (message.StepGrant.RunId == fastTimeRunId) {
                        var grant = new StepGrant(message.StepGrant, fastTimeRunId);
                        rti.DefaultDispatchMode = DispatchMode.Buffered; // switch to Buffered on first step
                        rti.FlushBuffers(); // dispatch messages buffered since last step
                        if (stepFn != null) {
                            try {
                                stepFn(grant);
                                CompleteStep(grant);
                            } catch (Exception e) {
                                CompleteStep(grant, failed: true, reason: e.Message);
                            }
                        } else {
                            OnStepGrant(grant);
                            grantQueue.Add(grant);
                        }
                    }
                    break;
            }
        }

        private void ResetFastTime() {
            if (fastTimeRunId != null) {
                fastTimeRunId = null;
                fastTimeControllerClientId = null;
                // Cancel any thread blocking in GetStepGrant() and drain the queue
                resetCts.Cancel();
                resetCts = new CancellationTokenSource();
                while (grantQueue.TryTake(out _)) {}
                rti.DefaultDispatchMode = DispatchMode.Immediate;
                rti.FlushBuffers();
                rti.FastTimeMode = false;
            }
        }

        private void Receive(RuntimeControl message) {
            switch (message.ControlCase) {
                case RuntimeControl.ControlOneofCase.Reset:
                    OnResetEndStop();
                    OnReset();
                    rti.State = RuntimeState.Initial;
                    if (fastTimeEnabled) ResetFastTime();
                    break;
                case RuntimeControl.ControlOneofCase.LoadScenario:
                    Scenario = null;
                    var playback = rti.State == RuntimeState.Playback;
                    rti.State = RuntimeState.Loading;
                    var success = OnLoadScenario(message.LoadScenario, playback);
                    if (!success) {
                        rti.State = RuntimeState.Unknown;
                        return;
                    }
                    Scenario = message.LoadScenario;
                    rti.Publish(RTIChannel.RuntimeControl, new RuntimeControl {
                        CurrentScenario = new RuntimeControl.Types.ScenarioSpecification { Name = Scenario.Name }
                    });
                    rti.State = playback ? RuntimeState.Playback : RuntimeState.Ready;
                    break;
                case RuntimeControl.ControlOneofCase.RequestCurrentScenario:
                    if (Scenario != null) {
                        rti.Publish(RTIChannel.RuntimeControl, new RuntimeControl {
                            CurrentScenario = new RuntimeControl.Types.ScenarioSpecification { Name = Scenario.Name }
                        });
                    }
                    break;
                case RuntimeControl.ControlOneofCase.Start:
                    OnStart();
                    rti.State = RuntimeState.Running;
                    break;
                case RuntimeControl.ControlOneofCase.Play:
                    OnPlay();
                    rti.State = RuntimeState.Playback;
                    if (fastTimeEnabled) ResetFastTime();
                    break;
                case RuntimeControl.ControlOneofCase.Pause:
                    OnPause();
                    if (rti.State == RuntimeState.Playback || rti.State == RuntimeState.PlaybackPaused)
                        rti.State = RuntimeState.PlaybackPaused;
                    else if (rti.State != RuntimeState.End && rti.State != RuntimeState.PlaybackEnd
                             && rti.State != RuntimeState.Stopped && rti.State != RuntimeState.PlaybackStopped)
                        rti.State = RuntimeState.Paused;
                    break;
                case RuntimeControl.ControlOneofCase.End:
                    OnResetEndStop();
                    OnEndStop();
                    OnEnd();
                    rti.State = rti.State == RuntimeState.Playback ? RuntimeState.PlaybackEnd : RuntimeState.End;
                    if (fastTimeEnabled) ResetFastTime();
                    break;
                case RuntimeControl.ControlOneofCase.Stop:
                    OnResetEndStop();
                    OnEndStop();
                    OnStop();
                    if (rti.State == RuntimeState.Playback || rti.State == RuntimeState.PlaybackPaused
                        || rti.State == RuntimeState.PlaybackStopped || rti.State == RuntimeState.PlaybackEnd)
                        rti.State = RuntimeState.PlaybackStopped;
                    else
                        rti.State = RuntimeState.Stopped;
                    if (fastTimeEnabled) ResetFastTime();
                    break;
                case RuntimeControl.ControlOneofCase.SetTimeScale:
                    TimeScale = message.SetTimeScale.TimeScale;
                    OnTimeScale(message.SetTimeScale.TimeScale);
                    break;
                case RuntimeControl.ControlOneofCase.TimeSync:
                    TimeScale = message.TimeSync.TimeScale;
                    OnTimeSync(message.TimeSync);
                    break;
                case RuntimeControl.ControlOneofCase.CurrentScenario:
                    Scenario = new RuntimeControl.Types.ScenarioSpecification { Name = message.CurrentScenario.Name };
                    break;
                case RuntimeControl.ControlOneofCase.Seek:
                    var prevState = rti.State;
                    OnSeek(message.Seek);
                    if (prevState == rti.State && rti.State != RuntimeState.Playback && rti.State != RuntimeState.Running && rti.State != RuntimeState.Paused) {
                        rti.State = RuntimeState.PlaybackPaused;
                    }
                    break;
            }
        }
    }
}
